File: internal/cloud/uploads.go
package cloud

import (
	"fmt"
	"io"
	"mime"
	"net/http"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"time"
)

func publicUpload(x object) object {
	result := selectFields(x, "id", "name", "size", "kind", "offset", "hashes", "result")
	result["chunkBytes"] = ChunkBytes
	return result
}
func (a *API) uploadLock(fn handler) handler {
	return func(w http.ResponseWriter, r *http.Request, s *Session) error {
		id := r.PathValue("id")
		a.mu.Lock()
		if a.uploads[id] {
			a.mu.Unlock()
			return fail(409, "Upload is busy. Retry this request.")
		}
		a.uploads[id] = true
		a.mu.Unlock()
		defer func() { a.mu.Lock(); delete(a.uploads, id); a.mu.Unlock() }()
		return fn(w, r, s)
	}
}
func (a *API) uploadDir(id string) string { return filepath.Join(a.Store.Directory, "uploads", id) }
func (a *API) installUploads() {
	a.handle("POST /api/uploads", func(w http.ResponseWriter, r *http.Request, s *Session) error {
		input, err := body(r)
		if err != nil {
			return err
		}
		name, kind := str(input["name"]), str(input["kind"])
		if input["kind"] == nil {
			kind = "media"
		}
		if name == "" || jsLength(name) > 255 || strings.IndexFunc(name, func(c rune) bool { return c < 32 }) >= 0 {
			return fail(400, "Invalid filename.")
		}
		size := int64(num(input["size"]))
		limit := a.options.MaxUpload
		if kind == "effect" && limit > 1024*1024*1024 {
			limit = 1024 * 1024 * 1024
		}
		if (kind != "media" && kind != "effect") || !integer(input["size"]) || size <= 0 || size > limit {
			return fail(413, "File exceeds the upload limit.")
		}
		item := object{"id": uuid(), "owner": s.Owner, "name": filepath.Base(name), "size": number(size), "kind": kind, "offset": float64(0), "hashes": []any{}, "createdAt": now()}
		err = a.Store.Transaction(r.Context(), func(tx *Store) error {
			used, e := tx.Usage(r.Context(), s.Owner)
			if e != nil {
				return e
			}
			if used+size > a.options.Quota {
				return fail(413, "Account storage quota exceeded. Remove unused files first.")
			}
			uploads, e := tx.List(r.Context(), "upload", s.Owner)
			if e != nil {
				return e
			}
			count := 0
			for _, x := range uploads {
				if x["result"] == nil {
					count++
				}
			}
			if count >= 8 {
				return fail(429, "Too many incomplete uploads.")
			}
			_, e = tx.Put(r.Context(), "upload", item, size)
			return e
		})
		if err != nil {
			return err
		}
		write(w, 201, publicUpload(item))
		return nil
	})
	a.handle("GET /api/uploads", func(w http.ResponseWriter, r *http.Request, s *Session) error {
		items, err := a.Store.List(r.Context(), "upload", s.Owner)
		if err != nil {
			return err
		}
		result := []object{}
		for _, item := range items {
			if item["result"] == nil {
				result = append(result, publicUpload(item))
			}
		}
		write(w, 200, result)
		return nil
	})
	a.handle("GET /api/uploads/{id}", func(w http.ResponseWriter, r *http.Request, s *Session) error {
		item, err := a.Store.Need(r.Context(), "upload", r.PathValue("id"), s.Owner)
		if err != nil {
			return err
		}
		write(w, 200, publicUpload(item))
		return nil
	})
	a.handle("PUT /api/uploads/{id}/chunks/{offset}", a.uploadLock(a.putChunk))
	a.handle("POST /api/uploads/{id}/complete", a.uploadLock(a.completeUpload))
	a.handle("DELETE /api/uploads/{id}", a.uploadLock(func(w http.ResponseWriter, r *http.Request, s *Session) error {
		item, err := a.Store.Need(r.Context(), "upload", r.PathValue("id"), s.Owner)
		if err != nil {
			return err
		}
		if item["result"] != nil {
			return fail(409, "Completed uploads are managed in the media library.")
		}
		if err = os.RemoveAll(a.uploadDir(str(item["id"]))); err != nil {
			return err
		}
		if err = a.Store.Remove(r.Context(), "upload", str(item["id"]), s.Owner); err != nil {
			return err
		}
		write(w, 200, object{"deleted": true})
		return nil
	}))
}
func (a *API) putChunk(w http.ResponseWriter, r *http.Request, s *Session) error {
	item, err := a.Store.Need(r.Context(), "upload", r.PathValue("id"), s.Owner)
	if err != nil {
		return err
	}
	content, _, e := mime.ParseMediaType(r.Header.Get("Content-Type"))
	if item["result"] != nil || e != nil || content != "application/octet-stream" {
		return fail(400, "Invalid upload chunk.")
	}
	offset, e := strconv.ParseInt(r.PathValue("offset"), 10, 64)
	if e != nil || offset < 0 || offset%ChunkBytes != 0 {
		return fail(400, "Invalid chunk offset.")
	}
	data, e := io.ReadAll(http.MaxBytesReader(w, r.Body, ChunkBytes))
	if e != nil {
		return e
	}
	if len(data) == 0 {
		return fail(400, "Invalid upload chunk.")
	}
	digest := hash(data)
	if r.Header.Get("X-Content-SHA256") != digest {
		return fail(400, "Chunk digest mismatch.")
	}
	hashes := array(item["hashes"])
	committed := int64(num(item["offset"]))
	size := int64(num(item["size"]))
	if offset < committed {
		index := int(offset / ChunkBytes)
		if index >= len(hashes) || hashes[index] != digest {
			return fail(409, "Previously uploaded chunk differs.")
		}
		write(w, 200, publicUpload(item))
		return nil
	}
	expected := size - offset
	if expected > ChunkBytes {
		expected = ChunkBytes
	}
	if offset != committed || int64(len(data)) != expected {
		return fail(409, "Resume from the committed offset.")
	}
	directory := a.uploadDir(str(item["id"]))
	if err = os.MkdirAll(directory, 0700); err != nil {
		return err
	}
	target := filepath.Join(directory, strconv.FormatInt(offset, 10))
	tmp := target + ".tmp"
	if err = durableFile(tmp, data); err != nil {
		return err
	}
	if err = os.Rename(tmp, target); err != nil {
		return err
	}
	if err = syncDir(directory); err != nil {
		return err
	}
	item["offset"] = number(offset + int64(len(data)))
	item["hashes"] = append(hashes, digest)
	if _, err = a.Store.Put(r.Context(), "upload", item, size); err != nil {
		return err
	}
	write(w, 200, publicUpload(item))
	return nil
}
func durableFile(file string, data []byte) error {
	f, err := os.OpenFile(file, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, 0600)
	if err != nil {
		return err
	}
	defer f.Close()
	if _, err = f.Write(data); err != nil {
		return err
	}
	return f.Sync()
}
func syncDir(directory string) error {
	f, err := os.Open(directory)
	if err != nil {
		return err
	}
	defer f.Close()
	return f.Sync()
}
func (a *API) completeUpload(w http.ResponseWriter, r *http.Request, s *Session) error {
	item, err := a.Store.Need(r.Context(), "upload", r.PathValue("id"), s.Owner)
	if err != nil {
		return err
	}
	if item["result"] != nil {
		write(w, 200, item["result"])
		return nil
	}
	if item["offset"] != item["size"] {
		return fail(409, "Upload is incomplete.")
	}
	directory := a.uploadDir(str(item["id"]))
	target := filepath.Join(directory, "source")
	f, err := os.OpenFile(target, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0600)
	if err != nil {
		return err
	}
	hashes := array(item["hashes"])
	size := int64(num(item["size"]))
	published := false
	defer func() {
		if !published {
			_ = os.Remove(target)
		}
	}()
	err = func() error {
		defer f.Close()
		for offset := int64(0); offset < size; offset += ChunkBytes {
			if err := r.Context().Err(); err != nil {
				return err
			}
			data, err := os.ReadFile(filepath.Join(directory, strconv.FormatInt(offset, 10)))
			if err != nil {
				return err
			}
			index := int(offset / ChunkBytes)
			if index >= len(hashes) || hash(data) != hashes[index] {
				return fail(409, "Stored chunk is damaged. Restart the upload.")
			}
			if _, err = f.Write(data); err != nil {
				return err
			}
		}
		return f.Sync()
	}()
	if err != nil {
		return err
	}
	result, err := a.rpc(r, object{"op": "inspect", "input": object{"file": target, "name": item["name"], "kind": item["kind"]}}, 30*time.Minute)
	if err != nil {
		return err
	}
	asset := obj(obj(result)["asset"])
	if str(asset["id"]) == "" {
		return fail(502, "Invalid media inspection result.")
	}
	asset["owner"] = s.Owner
	item["result"] = obj(result)["public"]
	err = a.Store.Transaction(r.Context(), func(tx *Store) error {
		if _, err := tx.Put(r.Context(), str(item["kind"]), asset, size); err != nil {
			return err
		}
		_, err := tx.Put(r.Context(), "upload", item, 0)
		return err
	})
	if err != nil {
		return err
	}
	published = true
	for offset := int64(0); offset < size; offset += ChunkBytes {
		if err = os.Remove(filepath.Join(directory, fmt.Sprint(offset))); err != nil && !os.IsNotExist(err) {
			return err
		}
	}
	write(w, 200, item["result"])
	return nil
}

File: server/cloud/uploads.mjs
import { randomUUID } from 'node:crypto';
import { mkdir, writeFile, rename, rm, open, readFile } from 'node:fs/promises';
import path from 'node:path';
import { digest, requireValue } from './store.mjs';
import { inspectMedia, publicMedia } from '../media.mjs';
import { inspectEffect, publicEffect } from '../effects.mjs';

export const CHUNK_BYTES = 8 * 1024 ** 2;
export function installUploads(app, route, { store, quota, maxUpload, signal }) {
  const locks = new Set();
  const folder = id => path.join(store.directory, 'uploads', id);
  const safe = item => ({ id: item.id, name: item.name, size: item.size, kind: item.kind, offset: item.offset, hashes: item.hashes, chunkBytes: CHUNK_BYTES, result: item.result });
  const locked = fn => route(async (req, res) => {
    const key = req.params.id;
    requireValue(!locks.has(key), 409, 'Upload is busy. Retry this request.');
    locks.add(key); try { await fn(req, res); } finally { locks.delete(key); }
  });
  app.post('/api/uploads', route(async (req, res) => {
    const { name, size, kind = 'media' } = req.body;
    requireValue(typeof name === 'string' && name.length > 0 && name.length <= 255 && !/[\x00-\x1f]/.test(name), 400, 'Invalid filename.');
    requireValue(['media', 'effect'].includes(kind) && Number.isSafeInteger(size) && size > 0 && size <= Math.min(maxUpload, kind === 'effect' ? 1024 ** 3 : maxUpload), 413, 'File exceeds the upload limit.');
    const upload = { id: randomUUID(), owner: req.session.owner, name: path.basename(name), size, kind, offset: 0, hashes: [], createdAt: Date.now() };
    store.transaction(() => {
      requireValue(store.usage(upload.owner) + size <= quota, 413, 'Account storage quota exceeded. Remove unused files first.');
      requireValue(store.list('upload', upload.owner).filter(x => !x.result).length < 8, 429, 'Too many incomplete uploads.');
      store.put('upload', upload, size);
    });
    res.status(201).json(safe(upload));
  }));
  app.get('/api/uploads', (req, res) => res.json(store.list('upload', req.session.owner).filter(x => !x.result).map(safe)));
  app.get('/api/uploads/:id', (req, res) => res.json(safe(store.need('upload', req.params.id, req.session.owner))));
  app.put('/api/uploads/:id/chunks/:offset', locked(async (req, res) => {
    const item = store.need('upload', req.params.id, req.session.owner), offset = Number(req.params.offset), bytes = req.body;
    requireValue(!item.result && Buffer.isBuffer(bytes) && bytes.length > 0 && bytes.length <= CHUNK_BYTES, 400, 'Invalid upload chunk.');
    requireValue(Number.isSafeInteger(offset) && offset >= 0 && offset % CHUNK_BYTES === 0, 400, 'Invalid chunk offset.');
    const hash = digest(bytes);
    requireValue(req.headers['x-content-sha256'] === hash, 400, 'Chunk digest mismatch.');
    if (offset < item.offset) { requireValue(item.hashes[offset / CHUNK_BYTES] === hash, 409, 'Previously uploaded chunk differs.'); res.json(safe(item)); return; }
    requireValue(offset === item.offset && bytes.length === Math.min(CHUNK_BYTES, item.size - offset), 409, 'Resume from the committed offset.');
    await mkdir(folder(item.id), { recursive: true, mode: 0o700 });
    const target = path.join(folder(item.id), String(offset)), temporary = `${target}.tmp`;
    // A chunk is committed only after its durable file exists. Orphaned files can be safely replaced on retry.
    const handle = await open(temporary, 'w', 0o600);
    try { await handle.writeFile(bytes); await handle.sync(); } finally { await handle.close(); }
    await rename(temporary, target);
    item.offset += bytes.length; item.hashes.push(hash); store.put('upload', item, item.size);
    res.json(safe(item));
  }));
  app.post('/api/uploads/:id/complete', locked(async (req, res) => {
    const item = store.need('upload', req.params.id, req.session.owner);
    if (item.result) { res.json(item.result); return; }
    requireValue(item.offset === item.size, 409, 'Upload is incomplete.');
    const target = path.join(folder(item.id), 'source'), handle = await open(target, 'w', 0o600);
    try {
      for (let offset = 0; offset < item.size; offset += CHUNK_BYTES) {
        signal.throwIfAborted();
        const bytes = await readFile(path.join(folder(item.id), String(offset)));
        requireValue(digest(bytes) === item.hashes[offset / CHUNK_BYTES], 409, 'Stored chunk is damaged. Restart the upload.');
        await handle.writeFile(bytes);
      }
      await handle.sync();
    } finally { await handle.close(); }
    let asset;
    try {
      const probe = await open(target, 'r'), header = Buffer.alloc(12);
      try { await probe.read(header, 0, 12, 0); } finally { await probe.close(); }
      const atom = header.toString('ascii', 4, 8), magic = header.toString('ascii', 0, 4);
      if (item.kind === 'media') {
        requireValue(['ftyp', 'moov', 'mdat', 'wide', 'free'].includes(atom), 400, 'Upload an MP4 or MOV file, not a playlist or URL.');
        asset = await inspectMedia(target, item.name, signal, { inputFormat: 'mov' });
        requireValue(asset.duration <= 7200 && asset.width > 0 && asset.height > 0 && asset.width * asset.height <= 3840 * 2160 && asset.fps <= 120 && asset.audioTracks.length <= 32 && asset.audioTracks.every(x => x.channels <= 8), 400, 'Cloud beta supports up to two hours, 4K, 120 fps and eight audio channels.');
      } else {
        requireValue(magic === 'RIFF' || magic === 'fLaC' || magic === 'OggS' || ['ftyp', 'moov'].includes(atom) || header.toString('ascii', 0, 3) === 'ID3' || (header[0] === 255 && (header[1] & 0xe0) === 0xe0), 400, 'Upload WAV, FLAC, Ogg, MP3, AAC or M4A audio, not a playlist.');
        asset = await inspectEffect(target, item.name, signal);
      }
    }
    catch (error) { await rm(target, { force: true }); throw error; }
    const result = (item.kind === 'media' ? publicMedia : publicEffect)(asset);
    store.transaction(() => {
      store.put(item.kind, { ...asset, owner: item.owner }, item.size);
      store.put('upload', { ...item, result }, 0);
    });
    for (let offset = 0; offset < item.size; offset += CHUNK_BYTES) await rm(path.join(folder(item.id), String(offset)), { force: true });
    res.json(result);
  }));
  app.delete('/api/uploads/:id', locked(async (req, res) => {
    const item = store.need('upload', req.params.id, req.session.owner);
    requireValue(!item.result, 409, 'Completed uploads are managed in the media library.');
    await rm(folder(item.id), { recursive: true, force: true }); store.remove('upload', item.id, item.owner); res.json({ deleted: true });
  }));
}
